### PR TITLE
Add HKDF provider

### PR DESCRIPTION
**Pull request by microshine**

### CODE REVIEW

File: src/hkdf/hkdf.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { KeyUsages } from "../../build/webcrypto-core";
+import { ProviderCrypto } from "../provider";
+import { BufferSourceConverter } from "../utils";
+
+export abstract class HkdfProvider extends ProviderCrypto {
+
+  public name = "HMAC";
```
Comment (LiraNuna) [inline src/hkdf/hkdf.ts]: HKDF?

Comment (microshine) [reply src/hkdf/hkdf.ts]: oops :)

Comment (microshine) [reply src/hkdf/hkdf.ts]: writing tests right now

Comment (LiraNuna) [reply src/hkdf/hkdf.ts]: Happens to the best of us

File: src/hkdf/hkdf.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  public checkImportKey(format: KeyFormat, keyData: JsonWebKey | ArrayBuffer, algorithm: Algorithm, extractable: boolean, keyUsages: KeyUsage[]) {
+    super.checkImportKey(format, keyData, algorithm, extractable, keyUsages);
+    if (extractable) {
+      // If extractable is not false, then throw a SyntaxError
+      throw new SyntaxError("extractable: Must be False");
+    }
+  }
+
+  public abstract onImportKey(format: KeyFormat, keyData: JsonWebKey | ArrayBuffer, algorithm: Algorithm, extractable: boolean, keyUsages: KeyUsage[]): Promise<CryptoKey>;
+  public abstract onDeriveBits(algorithm: Pbkdf2Params, baseKey: CryptoKey, length: number): Promise<ArrayBuffer>;
```
Comment (LiraNuna) [inline src/hkdf/hkdf.ts]: `Pbkdf2Params` should be `HkdfParams`

File: src/hkdf/hkdf.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  public hashAlgorithms = ["SHA-1", "SHA-256", "SHA-384", "SHA-512"];
+  public usages: KeyUsages = ["deriveKey", "deriveBits"];
+
+  public checkAlgorithmParams(algorithm: HkdfParams) {
+    // hash
+    this.checkRequiredProperty(algorithm, "hash");
+    this.checkHashAlgorithm(algorithm.hash as Algorithm, this.hashAlgorithms);
+
+    // salt
+    this.checkRequiredProperty(algorithm, "salt");
```
Comment (LiraNuna) [inline src/hkdf/hkdf.ts]: 'salt' is optional for HKDF

File: src/hkdf/hkdf.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if (!BufferSourceConverter.isBufferSource(algorithm.salt)) {
+      throw new TypeError("salt: Is not of type '(ArrayBuffer or ArrayBufferView)'");
+    }
+
+    // info
+    this.checkRequiredProperty(algorithm, "info");
```
Comment (LiraNuna) [inline src/hkdf/hkdf.ts]: 'info' is optional for HKDF

File: src/hkdf/hkdf.ts (hunk 3 above)
Comment (microshine) [reply src/hkdf/hkdf.ts]: Please check W3 spec
https://www.w3.org/TR/WebCryptoAPI/#dfn-HkdfParams

File: src/hkdf/hkdf.ts (hunk 4 above)
Comment (microshine) [reply src/hkdf/hkdf.ts]: https://www.w3.org/TR/WebCryptoAPI/#dfn-HkdfParams

File: src/hkdf/hkdf.ts (hunk 2 above)
Comment (microshine) [reply src/hkdf/hkdf.ts]: Thank you. I'll fix it

File: src/hkdf/hkdf.ts (hunk 3 above)
Comment (microshine) [reply src/hkdf/hkdf.ts]: Chrome throws error if `salt` or `info` params are missed
image

Comment (LiraNuna) [reply src/hkdf/hkdf.ts]: That is an odd choice by the W3C... The HKDF RFC says that both salt and info are optional and provides way to generate default values.

Sorry for the false comment, I wonder why the W3C chose not to follow the RFC.

**Pull request merged**